pkg/cli/admin/waitfornoderestart/wait_for_node_restart.go

```diff
@@ -0,0 +1,281 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	nodeKind = schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Node"}
+)
+
+const (
+	RemoveOldTrustFieldManager = "remove-old-trust"
+)
+
+type WaitForNodeRestartRuntime struct {
+	ResourceFinder genericclioptions.ResourceFinder
+	KubeClient     kubernetes.Interface
+	DynamicClient  dynamic.Interface
+
+	RebootNumber int
+
+	masterRestartNumber int
+	workerRestartNumber int
+
+	genericclioptions.IOStreams
+}
+
+func (r *WaitForNodeRestartRuntime) setRestartNumber(ctx context.Context) error {
+	if r.RebootNumber > 0 {
+		r.masterRestartNumber = r.RebootNumber
+		r.workerRestartNumber = r.RebootNumber
+		return nil
+	}
+
+	masterMachineConfig, err := r.DynamicClient.Resource(rebootmachineconfigpool.MachineConfigResource).Get(ctx, rebootmachineconfigpool.MasterRebootingMachineConfigName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+	// do nothing.  If we need to use it, we'll fail
+	case err != nil:
+		return fmt.Errorf("unable to read existing state: %w", err)
+	default:
+		r.masterRestartNumber, err = rebootmachineconfigpool.GetRestartNumber(masterMachineConfig)
+		if err != nil {
+			return fmt.Errorf("unable to parse master restart number: %w", err)
+		}
+	}
+
+	workerMachineConfig, err := r.DynamicClient.Resource(rebootmachineconfigpool.MachineConfigResource).Get(ctx, rebootmachineconfigpool.WorkerRebootingMachineConfigName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+	// do nothing.  If we need to use it, we'll fail
+	case err != nil:
+		return fmt.Errorf("unable to read existing state: %w", err)
+	default:
+		r.workerRestartNumber, err = rebootmachineconfigpool.GetRestartNumber(workerMachineConfig)
+		if err != nil {
+			return fmt.Errorf("unable to parse worker restart number: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *WaitForNodeRestartRuntime) Run(ctx context.Context) error {
+	if err := r.setRestartNumber(ctx); err != nil {
+		return err
+	}
+
+	hasMaster := false
+	hasWorker := false
+	remainingNodes := []*corev1.Node{}
+	visitor := r.ResourceFinder.Do()
+	err := visitor.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		if nodeKind != info.Object.GetObjectKind().GroupVersionKind() {
+			return fmt.Errorf("command must only be pointed at nodes")
+		}
+
+		uncastObj, ok := info.Object.(*unstructured.Unstructured)
+		if !ok {
+			return fmt.Errorf("not unstructured: %w", err)
+		}
+		node := &corev1.Node{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(uncastObj.Object, node); err != nil {
+			return fmt.Errorf("not a node: %w", err)
+		}
+		if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok {
+			hasMaster = true
+		} else {
+			hasWorker = true
```

Comment:
So I think the confusing part here might be that, technically, we are lumping all other roles into "worker". Like we discussed on the other PR, all custom pools will react to worker changes, so by initiating a worker pool reboot all custom pools will reboot in parallel.

Instead of describing the case as worker, maybe we just describe it as "worker + all custom pools" somehow. Same as to the reboot command, the user should be notified that the worker change will apply to all custom pools as well.

Reply:
Mostly naming. The main scenario I don't want to confuse people is, let's say they have `master, worker, infra` pools. They can't initiate a reboot just for infra, and worker reboots will also reboot infra.

Basically, make sure the user is aware of what's happening if there are non-worker pools